mixnet/topology/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+use std::{collections::HashMap, error::Error, net::SocketAddr};
+
+use nym_sphinx::addressing::nodes::NymNodeRoutingAddress;
+use rand::{rngs::OsRng, seq::IteratorRandom};
+use serde::{Deserialize, Serialize};
+use sphinx_packet::{crypto::PUBLIC_KEY_SIZE, route};
+
+pub type LayerId = u8;
+pub type MixnodeId = [u8; PUBLIC_KEY_SIZE];
+
+#[derive(Serialize, Deserialize, Clone, Debug, Default)]
+pub struct Topology {
+    pub layers: Vec<Layer>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct Layer {
+    pub id: LayerId,
+    pub nodes: HashMap<MixnodeId, Mixnode>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct Mixnode {
+    pub address: SocketAddr,
+    pub public_key: [u8; PUBLIC_KEY_SIZE],
+}
+
+impl Topology {
+    pub fn layer_id(&self, node_id: MixnodeId) -> Option<LayerId> {
+        for layer in self.layers.iter() {
+            if layer.nodes.contains_key(&node_id) {
+                return Some(layer.id);
+            }
+        }
+        None
+    }
+
+    pub fn random_route(&self, num_hops: usize) -> Result<Vec<route::Node>, Box<dyn Error>> {
+        if self.layers.len() < num_hops {
+            todo!("return error");
+        }
+
+        let mut route: Vec<route::Node> = Vec::new();
+
+        for layer_id in 0..num_hops {
+            let layer = self.layers.get(layer_id).unwrap();
+            route.push(
+                layer
+                    .random_node()
+                    .expect("layer is not empty")
+                    .clone()
+                    .try_into()
+                    .unwrap(),
+            );
+        }
+
+        Ok(route)
+    }
+}
+
+impl Layer {
+    pub fn random_node(&self) -> Option<&Mixnode> {
```

Comment:
Good point: 2fd055b.
I had to downgrade the `rand` and `rand_xoshiro` in the integration test, because `nym-sphinx` depends on the `rand@v0.7.3`. If we use the `rand@v0.8`, the following error occurs:
```rust
   --> mixnet/client/src/sender.rs:65:35
    |
65  |             .split_into_fragments(rng, plaintext_size_per_packet)
    |              -------------------- ^^^ the trait `rand_core::RngCore` is not implemented for `R`
```